core/src/main/scala/org/bitcoins/core/gcs/GCS.scala

```diff
@@ -0,0 +1,227 @@
+  */
+object GCS {
+
+  /**
+    * Given parameters and data, golomb-encodes the data
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#set-construction]]
+    */
+  def buildGCS(
+      data: Vector[ByteVector],
+      key: ByteVector,
+      p: UInt8,
+      m: UInt64): BitVector = {
+    val hashedValues = hashedSetConstruct(data, key, m)
+    val sortedHashedValues = hashedValues.sortWith(_ < _)
+    encodeSortedSet(sortedHashedValues, p)
+  }
+
+  /**
+    * Given parameters and data, constructs a GolombFilter for that data
+    */
+  def buildGolombFilter(
+      data: Vector[ByteVector],
+      key: ByteVector,
+      p: UInt8,
+      m: UInt64): GolombFilter = {
+    val encodedData = buildGCS(data, key, p, m)
+
+    GolombFilter(key, m, p, CompactSizeUInt(UInt64(data.length)), encodedData)
+  }
+
+  /**
+    * Given data, constructs a GolombFilter for that data using Basic Block Filter parameters
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#block-filters]]
+    */
+  def buildBasicBlockFilter(
+      data: Vector[ByteVector],
+      key: ByteVector): GolombFilter = {
+    buildGolombFilter(data, key, BlockFilter.P, BlockFilter.M)
+  }
+
+  private def sipHash(item: ByteVector, key: ByteVector): UInt64 = {
+    // https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#hashing-data-objects
+    val sipHashCParam = 2
+    val sipHashDParam = 4
+
+    val sh = new SipHash(sipHashCParam, sipHashDParam)
+
+    val keyParam = new KeyParameter(key.toArray)
+
+    sh.init(keyParam)
+
+    val offset = 0
+
+    sh.update(item.toArray, offset, item.length.toInt)
+
+    val digest = sh.doFinal()
+
+    UInt64.fromHex(digest.toHexString)
+  }
+
+  /**
+    * Hashes the item to the range [0, f)
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#hashing-data-objects]]
+    */
+  def hashToRange(item: ByteVector, f: UInt64, key: ByteVector): UInt64 = {
+    val hash = sipHash(item, key)
+
+    val bigInt = (hash.toBigInt * f.toBigInt) >> 64
+
+    UInt64(bigInt)
+  }
+
+  /**
+    * Hashes the items of a set of items
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#hashing-data-objects]]
+    */
+  def hashedSetConstruct(
+      rawItems: Vector[ByteVector],
+      key: ByteVector,
+      m: UInt64): Vector[UInt64] = {
+    val n = rawItems.length
+    val f = m * n
+
+    val hashedItemsBuilder = Vector.newBuilder[UInt64]
+
+    rawItems.foreach { item =>
+      val setValue = hashToRange(item, f, key)
+      hashedItemsBuilder.+=(setValue)
+    }
+
+    hashedItemsBuilder.result()
+  }
+
+  /**
+    * Converts num to unary (6 = 1111110)
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#golomb-rice-coding]]
+    *
+    * TODO: protect against large inputs which cause OutOfMemoryErrors and even larger ones which fail on toInt
```

Comment:
Has this been addressed? If not at minimum we should file an issue

Reply:
#528